exploration/0474-text-vs-code.md

```diff
@@ -103,7 +130,49 @@ equivalent of `\n`.
 
 ## Proposed Design
 
-TBD
+**Start in text mode**
+
+In this option, whitespace is trimmed from `pattern` constructs unless
+the pattern is quoted.
+
+```
+Hello world!
+
+Hello {$user}!
+
+{input $now :datetime dateStyle=long}
+Hello {$user}. Today is {$now}
+
+{local $now = {:systemGetCurrentTime :datetime dateStyle=medium}}
+Hello {$user}. Today is {$now}
+
+{match {$count :number integer=true}}
+{when 0}   Hello {$user}. Today is {$now} and you have no geese.
+{when one} Hello {$user}. Today is {$now} and you have {$count} goose.
+{when few} {  Hello {$user}, this message has spaces on the front and end.  }
+{when *}   Hello {$user}. Today is {$now} and you have {$count} geese.
+```
+
+Bear in mind that whitespace has no meaning in our syntax,
+so some of the above messages are actually:
+
+```
+{input $now :datetime dateStyle=long}Hello {$user}. Today is {$now}
+
+{match {$count :number option=value}}{when 0} Hello {$user}{when one} Hello {$user}{when *}{ Hello {$user} }
+```
+
+Key choices we should consider as a working group:
+
+- Do we like the double-brackets that `local` and `match` require?
+- Is the whitespace story clear? Will tools generate the extra `{}`
+  around patterns "just to be sure"?
```

Comment:
I think it would be clearer if the quotes were only around the leading and trailing whitespace. As proposed, it's unclear if `{ foo}` and `{foo }` are quoted patterns with significant whitespace, or unquoted literals without significant whitespace. So I think we'd need to drop unquoted literals if doing this.

I would be open to making the first character after `{` always significant, so that `{ }` would be a more ergonomic representation of an escaped space. This would not help with the confusion with unquoted literals, but it might make explicit quoted white space more palatable.

Reply:
I don't understand. An unquoted literal in a pattern is just text? Unquoted literals *never* contain whitespace (you have to quote any literal with internal or external whitespace).

I disagree with just quoting significant whitespace because it leads to even more brackety confusion. That is, when mentally comparing these:

```
{|  |}This has spaces.
{  This has spaces.}
```

... I think that they are both "valid", but that no one will write the top form (including because it produces a separate "part" and doesn't work neatly with translation.

I also am not keen on how it plays with other features of the syntax that cause the *need* for the pattern to be quoted at all (you only need to use `{}` when there is code present). I think all the curly brackets are hard to read:

```
{input $foo :function option=value}{|  |}This has spaces.
{input $foo :function option=value}{  This has spaces.}

{match {$foo}{when bar}{|  |}This has spaces.{when *}{  This has spaces.}
```

I am bothered the most by the `when` clauses because they descend into the same curly-bracketed demon-infested swamp that MF1 has.

Reply:
> I don't understand. An unquoted literal in a pattern is just text? Unquoted literals _never_ contain whitespace (you have to quote any literal with internal or external whitespace).

I meant either expanding the unquoted-literal syntax, or adding new syntax to make `{ }` parse as an escaped space.

> I disagree with just quoting significant whitespace because it leads to even more brackety confusion. That is, when mentally comparing these:
> 
> ```
> {|  |}This has spaces.
> {  This has spaces.}
> ```
> 
> ... I think that they are both "valid", but that no one will write the top form (including because it produces a separate "part" and doesn't work neatly with translation.

If you really do need those spaces, then should they not be separately quoted so that they'll be suitably protected during translation? I am clearly still not understanding the use cases for leading/trailing spaces that we ought to explicitly support with ergonomic syntax, rather than just making possible.

Reply:
> I meant either expanding the unquoted-literal syntax, or adding new syntax to make { } parse as an escaped space.

I don't think I'd support that.

> If you really do need those spaces, then should they not be separately quoted so that they'll be suitably protected during translation? I am clearly still not understanding the use cases for leading/trailing spaces that we ought to explicitly support with ergonomic syntax, rather than just making possible.

While it would be legal to quote spaces the way that you suggest (it already is, in our syntax), I don't think you would want to do that for the reasons I already cited.

I believe that the most common reason to want those spaces is to later concatenate strings. In that case, the spaces should be a normal part of the message's prose. You do not want to protect them from translation because many languages (Chinese, Japanese, Thai, etc...) will want to remove spaces in translation. Other types of whitespace, such as newlines or tabs, are also potentially open to needing translation, not just blind copying. And translation tools often do check that translated segments match exterior whitespace.

I don't think we're disagreeing unless what you're saying is that _quoted_ patterns (i.e. ones with `{`/`}` around them) do not preserve interior whitespace. Or is it your contention that should not just permit unquoted patterns but remove the ability to quote them?

Reply:
There are many good points brought by Addison and Stas, and I don't have much to add.
Only 2 points:
- We did spend a lot of time on this when we decided on the current approach. I didn't see anything that was not already considered at that time. Weirdness for first time users, the idea of trimming, already considered.
- For localisation WYSIWYG is best. As little escaping as possible.
  So `{   this is cleaner  }` than `{|    | than this |    |}`

Reply:
> @aphillips:
> > I meant either expanding the unquoted-literal syntax, or adding new syntax to make { } parse as an escaped space.
> 
> I don't think I'd support that.

Yeah, let's not do that.

> I don't think we're disagreeing unless what you're saying is that _quoted_ patterns (i.e. ones with `{`/`}` around them) do not preserve interior whitespace.

If we do end up with `{}` delimited patterns, whitespace immediately within them should be preserved.

> Or is it your contention that should not just permit unquoted patterns but remove the ability to quote them?

Yes. If we allow for simple undelimited patterns like `{$foo} is great`, then I think the readability of the syntax is significantly improved if all patterns are undelimited. As we do allow for whitespace to be quoted as a literal, and for `{||}` to effectively act as a pattern-start or pattern-end marker, I consider the cost of using `{}` not only for placeholders and statements but also patterns to be much greater than the benefit of providing yet another way of expressing a pattern with leading/trailing whitespace.

Reply:
I think we're pretty much in agreement on principles. Where I would maybe depart slightly is: I might be in favor of quoting patterns as a fallback. But much depends on the whitespace handling for that.